File: spring/database/mongo/src/main/java/ru/spliterash/spcore/database/mongo/repository/AbstractMongoSingletonRepository.java
package ru.spliterash.spcore.database.mongo.repository;

import lombok.AllArgsConstructor;
import org.springframework.core.ResolvableType;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Query;
import ru.spliterash.spcore.domain.repo.SingletonRepository;

import java.util.Optional;

@AllArgsConstructor
public abstract class AbstractMongoSingletonRepository<T> implements SingletonRepository<T> {
    private final MongoTemplate mongoTemplate;
    private final String collectionName;
    private final Class<T> documentClass;

    public AbstractMongoSingletonRepository(MongoTemplate mongoTemplate, String collectionName) {
        this.mongoTemplate = mongoTemplate;
        this.collectionName = collectionName;

        ResolvableType resolvableType = ResolvableType.forClass(AbstractMongoSingletonRepository.class, getClass());

        //noinspection unchecked
        documentClass = (Class<T>) resolvableType.getGeneric(0).getType();
    }

    @Override
    public Optional<T> get() {
        return Optional.ofNullable(mongoTemplate.findOne(new Query(), documentClass, collectionName));
    }

    @Override
    public void set(T t) {
        mongoTemplate.remove(new Query(), collectionName);
        mongoTemplate.save(t, collectionName);
    }
}

File: spring/database/mongo/src/main/java/ru/spliterash/spcore/database/mongo/repository/AbstractMongoSingletonRepository.kt
package ru.spliterash.spcore.database.mongo.repository

import kotlinx.coroutines.reactor.awaitSingle
import kotlinx.coroutines.reactor.awaitSingleOrNull
import org.springframework.core.ResolvableType
import org.springframework.data.mongodb.core.ReactiveMongoTemplate
import org.springframework.data.mongodb.core.query.Query
import ru.spliterash.spcore.domain.repo.SingletonRepository
import java.util.*

abstract class AbstractMongoSingletonRepository<T>(
    private val mongoTemplate: ReactiveMongoTemplate,
    private val collectionName: String
) : SingletonRepository<T> {
    private val documentClass: Class<T>

    init {
        val resolvableType = ResolvableType.forClass(AbstractMongoSingletonRepository::class.java, javaClass)
        @Suppress("UNCHECKED_CAST")
        documentClass = resolvableType.generics[0].type as Class<T>
    }

    override suspend fun get(): T? {
        return mongoTemplate.findOne(Query(), documentClass, collectionName).awaitSingleOrNull()
    }

    override suspend fun set(t: T) {
        mongoTemplate.remove(Query(), collectionName).awaitSingleOrNull()
        mongoTemplate.save(t, collectionName).awaitSingle()
    }
}